src/Core/Services/Implementations/EmergencyAccessService.cs

```diff
@@ -0,0 +1,307 @@
+
+            await _emergencyAccessRepository.ReplaceAsync(emergencyAccess);
+            await _mailService.SendEmergencyAccessAcceptedEmailAsync(granteeEmail, grantor.Email);
+
+            return emergencyAccess;
+        }
+
+        public async Task DeleteAsync(Guid emergencyAccessId, Guid grantorId)
+        {
+            var emergencyAccess = await _emergencyAccessRepository.GetByIdAsync(emergencyAccessId);
+            if (emergencyAccess == null || (emergencyAccess.GrantorId != grantorId && emergencyAccess.GranteeId != grantorId))
+            {
+                throw new BadRequestException("Emergency Access not valid.");
+            }
+
+            await _emergencyAccessRepository.DeleteAsync(emergencyAccess);
+        }
+
+        public async Task<EmergencyAccess> ConfirmUserAsync(Guid emergencyAcccessId, string key, Guid confirmingUserId)
+        {
+            var emergencyAccess = await _emergencyAccessRepository.GetByIdAsync(emergencyAcccessId);
+            if (emergencyAccess == null || emergencyAccess.Status != EmergencyAccessStatusType.Accepted ||
+                emergencyAccess.GrantorId != confirmingUserId)
+            {
+                throw new BadRequestException("Emergency Access not valid.");
+            }
+
+            var grantor = await _userRepository.GetByIdAsync(confirmingUserId);
+            var grantee = await _userRepository.GetByIdAsync(emergencyAccess.GranteeId.Value);
+
+            emergencyAccess.Status = EmergencyAccessStatusType.Confirmed;
+            emergencyAccess.KeyEncrypted = key;
+            emergencyAccess.Email = null;
+            await _emergencyAccessRepository.ReplaceAsync(emergencyAccess);
+            await _mailService.SendEmergencyAccessConfirmedEmailAsync(grantor.Name, grantee.Email);
+
+            return emergencyAccess;
+        }
+
+        public async Task SaveAsync(EmergencyAccess emergencyAccess, Guid savingUserId)
+        {
+            if (emergencyAccess.GrantorId != savingUserId)
+            {
+                throw new BadRequestException("Emergency Access not valid.");
+            }
+            
+            await _emergencyAccessRepository.ReplaceAsync(emergencyAccess);
+        }
+
+        public async Task InitiateAsync(Guid id, User initiatingUser)
+        {
+            var emergencyAccess = await _emergencyAccessRepository.GetByIdAsync(id);
+
+            if (emergencyAccess == null || emergencyAccess.GranteeId != initiatingUser.Id ||
+                emergencyAccess.Status != EmergencyAccessStatusType.Confirmed)
+            {
+                throw new BadRequestException("Emergency Access not valid.");
+            }
+
+            emergencyAccess.Status = EmergencyAccessStatusType.RecoveryInitiated;
+            emergencyAccess.RevisionDate = DateTime.UtcNow;
+            emergencyAccess.RecoveryInitiatedAt = DateTime.UtcNow;
+            emergencyAccess.LastNotificationAt = DateTime.UtcNow;
```

Comment:
create a single `now` variable so that all of these end up being the same value

Reply:
Whops, good catch.